File: backend/app/routes/projetos_adicionais.py
"""
Rotas /projetos-adicionais — CRUD completo de projetos adicionais por cliente.
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from app.services.storage_service import projetos_adicionais_storage

router = APIRouter()


class ProjetoCreate(BaseModel):
    cliente: str
    tipo: str = "lp"           # lp | apresentacao | site | comercial | outro
    nome: str
    valor: float = 0.0
    competencia: Optional[str] = None
    data_vencimento: Optional[str] = None
    status_pagamento: str = "pendente"  # pendente | pago | em_aberto | parcial
    data_recebimento: Optional[str] = None
    observacoes: Optional[str] = None


class ProjetoUpdate(BaseModel):
    cliente: Optional[str] = None
    tipo: Optional[str] = None
    nome: Optional[str] = None
    valor: Optional[float] = None
    competencia: Optional[str] = None
    data_vencimento: Optional[str] = None
    status_pagamento: Optional[str] = None
    data_recebimento: Optional[str] = None
    observacoes: Optional[str] = None


@router.get("")
def list_projetos():
    projetos = projetos_adicionais_storage.all()
    total        = sum(float(p.get("valor", 0)) for p in projetos)
    total_pago   = sum(float(p.get("valor", 0)) for p in projetos if p.get("status_pagamento") == "pago")
    total_pend   = sum(float(p.get("valor", 0)) for p in projetos if p.get("status_pagamento") != "pago")
    return {
        "projetos":       projetos,
        "total":          round(total, 2),
        "total_pago":     round(total_pago, 2),
        "total_pendente": round(total_pend, 2),
        "quantidade":     len(projetos),
        "pagos":          len([p for p in projetos if p.get("status_pagamento") == "pago"]),
        "pendentes":      len([p for p in projetos if p.get("status_pagamento") != "pago"]),
    }


@router.post("")
def create_projeto(body: ProjetoCreate):
    return projetos_adicionais_storage.create(body.model_dump())


@router.put("/{id}")
def update_projeto(id: str, body: ProjetoUpdate):
    payload = {k: v for k, v in body.model_dump().items() if v is not None}
    updated = projetos_adicionais_storage.update(id, payload)
    if not updated:
        raise HTTPException(status_code=404, detail="Projeto não encontrado")
    return updated


@router.delete("/{id}")
def delete_projeto(id: str):
    if not projetos_adicionais_storage.delete(id):
        raise HTTPException(status_code=404, detail="Projeto não encontrado")
    return {"ok": True}

File: functions/src/routes/projetosAdicionais.ts
import { Router } from "express";
import { projetosAdicionaisStorage } from "../lib/firestore";

const router = Router();

function toFloat(v: unknown): number {
  const n = parseFloat(String(v ?? "0"));
  return isFinite(n) ? n : 0;
}

// GET /projetos-adicionais
router.get("/", async (_req, res) => {
  try {
    const items = await projetosAdicionaisStorage.all();
    const pagos = items.filter((p) => p.status === "pago");
    const pendentes = items.filter((p) => p.status !== "pago");
    const total = items.reduce((s, p) => s + toFloat(p.valor), 0);
    const total_pago = pagos.reduce((s, p) => s + toFloat(p.valor), 0);
    const total_pendente = pendentes.reduce((s, p) => s + toFloat(p.valor), 0);
    const r = (n: number) => Math.round(n * 100) / 100;
    res.json({
      projetos: items,
      total: r(total), total_pago: r(total_pago), total_pendente: r(total_pendente),
      quantidade: items.length, pagos: pagos.length, pendentes: pendentes.length,
    });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// POST /projetos-adicionais
router.post("/", async (req, res) => {
  try {
    res.status(201).json(await projetosAdicionaisStorage.create(req.body));
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// PUT /projetos-adicionais/:id
router.put("/:id", async (req, res) => {
  try {
    const updated = await projetosAdicionaisStorage.update(req.params.id, req.body);
    if (!updated) { res.status(404).json({ detail: "Projeto não encontrado" }); return; }
    res.json(updated);
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// DELETE /projetos-adicionais/:id
router.delete("/:id", async (req, res) => {
  try {
    if (!await projetosAdicionaisStorage.delete(req.params.id)) { res.status(404).json({ detail: "Projeto não encontrado" }); return; }
    res.json({ ok: true });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

export default router;
